src/loader/init.c

```diff
@@ -0,0 +1,149 @@
+	if (guc_disable_load)
+		return;
+
+	/* Don't do a load if setting timescaledb.disable_load or doing an update */
+	if (query->commandType == CMD_UTILITY)
+	{
+		if (IsA(query->utilityStmt, VariableSetStmt))
+		{
+			VariableSetStmt *stmt = (VariableSetStmt *) query->utilityStmt;
+
+			if (strcmp(stmt->name, GUC_DISABLE_LOAD_NAME) == 0)
+			{
+				return;
+			}
+		}
+		else if (IsA(query->utilityStmt, AlterExtensionStmt))
+		{
+			AlterExtensionStmt *stmt = (AlterExtensionStmt *) query->utilityStmt;
+
+			if (strcmp(stmt->extname, EXTENSION_NAME) == 0)
+			{
+				if (extension_loaded())
+				{
+					ereport(ERROR,
+							(errmsg("Cannot update the extension after the old version has already been loaded"),
+							 errhint("You should start a new session and execute ALTER EXTENSION as the first command")));
+				}
+
+				return;
+			}
+		}
+	}
+	extension_check();
+}
+
+void
+_PG_init(void)
+{
+	if (!process_shared_preload_libraries_in_progress)
+	{
+		/* cannot use GUC variable here since extension not yet loaded */
+		char	   *allow_install_without_preload = GetConfigOptionByName("timescaledb.allow_install_without_preload", NULL, true);
+
+		if (allow_install_without_preload == NULL ||
+			strlen(allow_install_without_preload) != 2 ||
+			strncmp(allow_install_without_preload, "on", 2) != 0)
+		{
+			char	   *config_file = GetConfigOptionByName("config_file", NULL, false);
+
+			ereport(ERROR,
+					(errmsg("The timescaledb library is not preloaded"),
+					 errhint("Please preload the timescaledb library via shared_preload_libraries.\n\n"
+					 "This can be done by editing the config file at: %1$s\n"
+							 "and adding 'timescaledb' to the list in the shared_preload_libraries config.\n"
+							 "	# Modify postgresql.conf:\n	shared_preload_libraries = 'timescaledb'\n\n"
+							 "Another way to do this, if not preloading other libraries, is with the command:\n"
+							 "	echo \"shared_preload_libraries = 'timescaledb'\" >> %1$s \n\n"
+							 "(Will require a database restart.)\n\n"
+							 "If you REALLY know what you are doing and would like to load the library without preloading, you can disable this check with: \n"
+							 "	SET timescaledb.allow_install_without_preload = 'on';", config_file)));
+			return;
+		}
+	}
+	elog(INFO, "timescaledb loaded");
+
+	/* This is a safety-valve variable to prevent loading the full extension */
+	DefineCustomBoolVariable(GUC_DISABLE_LOAD_NAME, "Disable the loading of the actual extension",
+							 NULL,
+							 &guc_disable_load,
+							 false,
+							 PGC_USERSET,
+							 0,
+							 NULL,
+							 NULL,
+							 NULL);
+
+	/* cannot check for extension here since not inside a transaction yet */
+
+	CacheRegisterRelcacheCallback(inval_cache_callback, PointerGetDatum(NULL));
+
+	/*
+	 * using the post_parse_analyze_hook since it's the earliest available
+	 * hook
+	 */
+	post_parse_analyze_hook = post_analyze_hook;
+}
+
+void
+_PG_fini(void)
+{
+	/* No way to unregister relcache callback */
```

Comment:
Shouldn't we reset the `post_analyze_hook` here, especially since we allow loading this lib without preloading and to be future proof in case PostgreSQL ever allows unloading libs?